projects/hathor-wallet-mobile/0003-token-swap/token-swap.md

```diff
@@ -0,0 +1,282 @@
+Once the user confirms we can swap the tokens using the Dozer protocol.
+
+## Dozer token swap protocol
+
+The Dozer protocol for token swap is implemented as a nano contract on Hathor. 
+The methods being called for a token swap depend on the user input and the exchange expected path returned by the contract.
+
+The following chart should illustrate the decision making process to call each method.
+```mermaid
+flowchart TD
+
+    START(User defines tokens for the swap) --> ChooseValue{User writes in top or botton area}
+
+    %% Writing in the upper input will set forth a swap where you deposit the amount specified
+    ChooseValue -->|Top| InputToken[call 'find_best_swap_path']
+    InputToken -->|User confirms swap| InputPathLength{Swap path length?}
+    InputPathLength -->|1| InputSwap0Len[call 'swap_exact_tokens_for_tokens']
+    InputPathLength -->|>1| InputSwapNLen[call 'swap_exact_tokens_for_tokens_through_path']
+
+    %% Writing in the lower input will set forth a swap where you withdraw the amount specified
+    ChooseValue -->|Bottom| OutputToken[call 'find_best_swap_path_exact_output']
+    OutputToken -->|User confirms swap| OutputPathLength{Swap path length?}
+    OutputPathLength -->|1| OutputSwap0Len[call 'swap_tokens_for_exact_tokens']
+    OutputPathLength -->|>1| OutputSwapNLen[call 'swap_tokens_for_exact_tokens_through_path']
+
+    %% User changing the value should call find_best_swap method again
+    InputToken -->|User changed value| ChooseValue
+    OutputToken -->|User changed value| ChooseValue
+
+    %% Sending the method call should end the swap    
+    OutputSwapNLen --> END(Confirm NanoContract execution)
+    OutputSwap0Len --> END
+    InputSwap0Len --> END
+    InputSwapNLen --> END
+```
+
+The `find_best_swap*` methods return a tuple with the information:
+- `path`: Comma-separated string of swaps to be made (along which pools the swap will traverse) in the format `{tokenA}/{tokenB}/{fee1},{tokenB}/{tokenC}/{fee2}`
+- `amounts`: List of expected amounts at each step
+- `amount_out`/`amount_in`: Final expected output amount or expected deposit
+- `price_impact`: Overall price impact
+
+The swap information can be used so the user can review the swap and proceed to call the actual swap method.
+There are 2 swap methods depending on the number of itens on the calculated `path` if there is only 1 swap or multiple swaps.
+The `_through_path` variant should be called when the `path` length is greater than 1.
+
+### Slippage
+
+The actual exchange rate of a token swap is defined during the swap, but on Hathor you have to provide the deposit and withdrawal before running the method.
+To make this possible we have to account for slippage when building the transaction, to achieve this it depends if the known amount is the input token or output token.
+
+This means that the accepted slippage will always be defined by the user's wallet and the swap will automatically fail if slippage is more than expected by the user.

```

Comment:
I didn't understand this. How will the wallet define the slippage? I thought it would come from the contract